ios/FluentUI/Core/Theme/Tokens/TokenResolver.swift

```diff
@@ -0,0 +1,68 @@
+//
+//  Copyright (c) Microsoft Corporation. All rights reserved.
+//  Licensed under the MIT License.
+//
+
+import SwiftUI
+
+/// Contains all possible token overrides for a given control, and returns the appropriate token value for a given property.
+public class TokenResolver<ControlType: TokenizedControl>: NSObject, ObservableObject {
+
+    /// Returns the appropriate token value for a given key path on this control's token set.
+    ///
+    /// This method looks for the first value that does not match the default value provided by `defaultTokens`.
+    /// Priority order for tokens are `overrideTokens`, `themeTokens`, then finally `defaultTokens`.
+    func value<ValueType: Equatable>(_ keyPath: KeyPath<TokenType, ValueType>) -> ValueType {
+        let defaultValue = defaultTokens[keyPath: keyPath]
+        if let overrideValue = overrideTokens?[keyPath: keyPath], overrideValue != defaultValue {
```

Comment:
Do we have to compare it with the default value?

I'm thinking about this scenario:
- The default value of a given token is the blue color.
- There is a theme override changing that token to a green color.
- A specific instance of the control wants to have that token set back to the original blue value.

Wouldn't this scenario end up incorrectly returning green?

Reply:
Yep--this is totally a limitation that @huwilkes, @edjamesmsft, and I discovered. We weren't able to come up with a solution, so we decided to push forward with this approach anyway (it was the best we had), but I would LOVE to find out that you have an idea of how to solve this?

Things we considered:
* Separate out the token base class from the "default" tokens, so that the base can return a "null" value to be compared against
  * Problem: we double the number of class instances we have to ship for all controls
* Switch tokens from overriding a base class to implementing an `@objc` protocol with optional methods
  * Problem: we still need to have both a protocol and an implementing base class, which may introduce developer error in keeping them in sync (as well as still some binary bloat)

@huwilkes, @edjamesmsft, any others?

Reply:
I think we could potentially explore a hybrid of the 2 options that you mentioned:

What if instead of Base classes, we provide protocols with nullable properties and provide default implementations for the protocols that return nil?

That way we wouldn't need to expose the "default value class" publicly. Instead we would just expose the protocol (which could be even cheaper in terms of binary size).

That way the TokenResolver will be solely responsible for the order of precedence of the token values (instead of relying on inheritance).

Reply:
I like the overall approach you're suggesting, but unfortunately there are a few problems with presenting only a pure protocol to clients. In particular, that leaves them with no access to the current context's global and alias tokens, which are currently made available via the base `ControlTokens` class. Though I guess if we required `overrideTokens` to both conform to the given control's tokens, _and_ inherit from the base `ControlTokens` class, then I guess we could make it work...!

Reply:
Update: After verifying that this approach is technically possible, we still run into binary size issues--my initial conversion for `CardNudge` added another ~40kB, which is not feasible. So that approach is out.